ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,332 @@
+	HTTPStatusCode uint32 `json:"HTTP_status_code,omitempty"`
+	Type           string `json:"type,omitempty"`
+	DBType         string `json:"DB_type,omitempty"`
+	Hits           uint64 `json:"hits,omitempty"`
+	Errors         uint64 `json:"errors,omitempty"`
+	Duration       uint64 `json:"duration,omitempty"`
+	OkSummary      []byte `json:"okSummary,omitempty"`
+	ErrorSummary   []byte `json:"errorSummary,omitempty"`
+	Synthetics     bool   `json:"synthetics,omitempty"`
+	TopLevelHits   uint64 `json:"topLevelHits,omitempty"`
+}
+
+type spanSummary struct {
+	Start, Duration int64
+	Name            string
+	Resource        string
+	Service         string
+	Type            string
+	Error           int32
+	Hostname        string
+	Synthetics      bool
+	Env             string
+	StatusCode      uint32
+	Version         string
+	TopLevel        bool
+	Weight          float64
+}
+
+type concentrator struct {
+	In chan *spanSummary
+
+	mu      sync.Mutex
+	buckets map[int64]*rawBucket // buckets used to aggregate stats per timestamp
+
+	bufferLen int
+	oldestTs  int64
+	stop      chan struct{}
+	stopped   uint32 // atomic bool
+	cfg       *config
+}
+
+// defaultBufferLen represents the default buffer length; the number of bucket size
+// units used by the concentrator.
+const defaultBufferLen = 2
+
+func newConcentrator(c *config) *concentrator {
+	return &concentrator{
+		In:        make(chan *spanSummary, 10000),
+		bufferLen: defaultBufferLen,
+		stop:      make(chan struct{}),
+		oldestTs:  alignTs(time.Now().UnixNano()),
+		buckets:   make(map[int64]*rawBucket),
+		cfg:       c,
+	}
+}
+
+var bucketSize = (10 * time.Second).Nanoseconds()
+
+// alignTs returns the provided timestamp truncated to the bucket size.
+// It gives us the start time of the time bucket in which such timestamp falls.
+func alignTs(ts int64) int64 { return ts - ts%bucketSize }
+
+func (c *concentrator) Start() {
+	defer close(c.stop)
+	tick := time.NewTicker(time.Duration(bucketSize) * time.Nanosecond)
+	defer tick.Stop()
+	go func() {
+		for {
+			select {
+			case now := <-tick.C:
+				_ = now
+				fmt.Printf("\n\nFLUSHING: %#v\n\n", c.flush(now))
+			case <-c.stop:
+				return
+			}
+		}
+	}()
+	for {
+		select {
+		case ss := <-c.In:
+			btime := alignTs(ss.Start + ss.Duration)
+			if btime < c.oldestTs {
+				btime = c.oldestTs
+			}
+			b, ok := c.buckets[btime]
+			if !ok {
+				b = newRawBucket(uint64(btime))
+				c.buckets[btime] = b
+			}
+			b.handleSpan(ss)
+		case <-c.stop:
+			atomic.AddUint32(&c.stopped, 1)
+			return
+		}
+	}
+}
+
+func (c *concentrator) Stop() {
+	if atomic.LoadUint32(&c.stopped) > 0 {
+		return
+	}
+	c.stop <- struct{}{}
+	<-c.stop
+}
+
+func (c *concentrator) flush(timenow time.Time) statsPayload {
+	now := timenow.UnixNano()
+	sp := statsPayload{
+		Hostname: c.cfg.hostname,
+		Env:      c.cfg.env,
+		Version:  c.cfg.version,
+		Stats:    make([]statsBucket, 0, len(c.buckets)),
+	}
+	c.mu.Lock()
+	for ts, srb := range c.buckets {
```

Comment:
won't this flush even the current bucket?

Reply:
Fixed.